packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,436 @@
+ *   addItem: (vatID: string, item: string) => void,
+ *   readSpan: (vatID: string, startPos?: number) => Iterable<string>,
+ *   exportSpan: (vatID: string, startPos: number) => AsyncIterable<Uint8Array>
+ *   importSpan: (artifactName: string, exporter: SwingStoreExporter, artifactMetadata: Map) => Promise<void>,
+ *   getExportRecords: (includeHistorical: boolean) => Iterable<[key: string, value: string]>,
+ *   getArtifactNames: (includeHistorical: boolean) => Iterable<string>,
+ * }} TranscriptStore
+ *
+ * @typedef {{
+ *   dumpTranscripts: (includeHistorical?: boolean) => any,
+ * }} TranscriptStoreDebug
+ *
+ */
+
+function* empty() {
+  // Yield nothing
+}
+
+/**
+ * @param {unknown} position
+ * @returns {asserts position is number}
+ */
+
+function insistTranscriptPosition(position) {
+  assert.typeof(position, 'number');
+  assert(position >= 0);
+}
+
+/**
+ * @param {*} db
+ * @param {() => void} ensureTxn
+ * @param {(key: string, value: string) => void} noteExport
+ * @returns { TranscriptStore & TranscriptStoreDebug }
+ */
+export function makeTranscriptStore(db, ensureTxn, noteExport) {
+  db.exec(`
+    CREATE TABLE IF NOT EXISTS transcriptItems (
+      vatID TEXT,
+      position INTEGER,
+      item TEXT,
+      PRIMARY KEY (vatID, position)
+    )
+  `);
+
+  db.exec(`
+    CREATE TABLE IF NOT EXISTS transcriptSpans (
+      vatID TEXT,
+      startPos INTEGER,
+      endPos INTEGER,
+      hash TEXT,
+      current INTEGER,
+      size INTEGER,
+      PRIMARY KEY (vatID, startPos)
+    )
+  `);
+
+  const sqlDumpItemsQuery = db.prepare(`
+    SELECT vatID, position, item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY vatID, position
+  `);
+
+  const sqlDumpSpansQuery = db.prepare(`
+    SELECT vatID, startPos, endPos, current
+    FROM transcriptSpans
+    ORDER BY vatID, startPos
+  `);
+
+  function dumpTranscripts(includeHistorical = true) {
+    // debug function to return: dump[vatID][position] = item
+    const transcripts = {};
+    for (const spanRow of sqlDumpSpansQuery.iterate()) {
+      if (includeHistorical || spanRow.current) {
+        for (const row of sqlDumpItemsQuery.iterate(
+          spanRow.vatID,
+          spanRow.startPos,
+          spanRow.endPos,
+        )) {
+          const { vatID, position, item } = row;
+          if (!transcripts[vatID]) {
+            transcripts[vatID] = {};
+          }
+          transcripts[vatID][position] = item;
+        }
+      }
+    }
+    return transcripts;
+  }
+
+  /**
+   * Compute the hash for a transcript item.  This is computed by hashing
+   * together the hash from the previous item in its span together with its own text.
+   *
+   * @param {string} priorHash  The previous item's hash
+   * @param {string} item  The item itself
+   *
+   * @returns {string}  The hash of the combined parameters.
+   */
+  function computeItemHash(priorHash, item) {
```

Comment:
I went with `computeItemHash` because there's an individual (cumulative) hash field in each item and this function is what computes that, but it's actually an interesting question since although the hash is computed on a per-item basis item-by-item it is used for purposes of validating the span (albeit by checking item by item).  I'm inclined to leave the name as is since it seems clearer to me, but I'm not passionate about the design choice.

Reply:
> there's an individual (cumulative) hash field in each item

Wait, what? The transcript item itself doesn't store the hash, only the span does, or did I miss something?

Reply:
No you didn't miss something; I misspoke on account of conflating my mental model of what is going on with what is actually going on.  The items indeed don't contain individual hashes, because the only hash that matters for validation purposes is the hash produced by the last item in the sequence.  However, the hash itself is still computed incrementally on a per-item basis each time an item is added (or checked).  Like I said, I'm not passionate about the terminology choice here.